### PR TITLE
Feature HRNet

### PR DESCRIPTION
**Pull request by VladislavPatrushev**

### CODE REVIEW

File: src/models/heads/segmentation/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from . import hrnet_head
```
Comment (PososikTeam) [inline src/models/heads/segmentation/__init__.py:1]: empty line

File: src/models/backbones/utils/helpers.py [hunk 2]
```diff
@@ -23,7 +24,7 @@ def build_model_with_cfg(model_cls: Callable,
         default_cfg: model's default pretrained/task config
         **model_args: model args passed through to model __init__
```
Comment (PososikTeam) [inline src/models/backbones/utils/helpers.py:25]: documentation about model_cfg

File: src/models/heads/segmentation/hrnet_head.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+@HEADS.register_class
+class HRNetSegmentationHead(AbstractHead):
+    """HRNet head for segmentation tasks."""
+    def __init__(self, in_features: int, num_classes: int, kernel_size: int = 1):
+        """Init HRNetSegmentationHead.
+
+        Args:
+            in_features: Size of each input sample.
+            num_classes: Number of classes.
+            kernel_size: Kernel size.
```
Comment (PososikTeam) [inline src/models/heads/segmentation/hrnet_head.py:22]: If we have cases when kernel_size is not 1?

File: src/models/necks/__init__.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+from . import classification
+from . import segmentation
```
Comment (PososikTeam) [inline src/models/necks/__init__.py:2]: empty line

File: src/models/necks/classification/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from . import hrnet_classificaton_neck
```
Comment (PososikTeam) [inline src/models/necks/classification/__init__.py:1]: empty line

File: src/models/necks/classification/hrnet_classificaton_neck.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+class HRNetClassificationNeck(BaseModel):
+    """HRNet neck for classification task."""
+    def __init__(self, in_channels):
+        """Init HRNetClassificationNeck.
+
+        Args:
+            in_channels: Input channels.
+        """
+        super().__init__()
+        self.num_features = 2048
```
Comment (PososikTeam) [inline src/models/necks/classification/hrnet_classificaton_neck.py:22]: Is it always be 2048?

File: tests/models/backbones/test_hrnet.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+import unittest
```
Comment (PososikTeam) [inline tests/models/backbones/test_hrnet.py:1]: Is it convert to ONNX?

File: src/models/backbones/hrnet.py [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+        super().__init__()
+
+        self.num_inchannels = num_inchannels
+        self.num_branches = num_branches
+
+        self.branches = self._make_branches(num_branches, blocks, num_blocks, num_outchannels)
+        self.fuse_layers = self._make_fuse_layers()
+        self.fuse_act = nn.ReLU()
+
+    def _make_one_branch(self, branch_index, block, num_blocks, num_channels, stride=1):
```
Comment (PososikTeam) [inline src/models/backbones/hrnet.py:365]: Pltase describe all parameters and write documentation

File: src/models/backbones/hrnet.py [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+        return nn.Sequential(*layers)
+
+    def _make_branches(self, num_branches, block, num_blocks, num_channels):
+        branches = []
+        for i in range(num_branches):
+            branches.append(self._make_one_branch(i, block, num_blocks, num_channels))
+
+        return nn.ModuleList(branches)
+
+    def _make_fuse_layers(self):
```
Comment (PososikTeam) [inline src/models/backbones/hrnet.py:390]: Please describe all parameters and write documentation

File: src/models/backbones/hrnet.py [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+    def __init_weights(self):
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                nn.init.kaiming_normal_(
+                    m.weight, mode='fan_out', nonlinearity='relu')
+            elif isinstance(m, nn.BatchNorm2d):
+                nn.init.constant_(m.weight, 1)
+                nn.init.constant_(m.bias, 0)
+
+    def __make_transition_layer(self, num_channels_pre_layer, num_channels_cur_layer):
```
Comment (PososikTeam) [inline src/models/backbones/hrnet.py:529]: Please describe all parameters and write documentation

File: src/models/backbones/hrnet.py [hunk 11]
```diff
@@ -0,0 +1,696 @@
...
+                                   act_layer=None)
+
+        layers = [block(inplanes, planes, stride, downsample)]
+        inplanes = planes * block.expansion
+        for i in range(1, blocks):
+            layers.append(block(inplanes, planes))
+
+        return nn.Sequential(*layers)
+
+    def __make_stage(self, layer_config, num_inchannels):
```
Comment (PososikTeam) [inline src/models/backbones/hrnet.py:581]: Please describe all parameters and write documentation

File: src/models/backbones/hrnet.py [hunk 12]
```diff
@@ -0,0 +1,696 @@
...
+            x: Input tensor.
+        """
+        x = self.convbnact1(x)
+        x = self.convbnact2(x)
+
+        yl = self.__stages(x)
+
+        return yl
+
+    def forward_backbone_features(self, x: Tensor) -> Tuple[List[Tensor], List[Tensor]]:
```
Comment (PososikTeam) [inline src/models/backbones/hrnet.py:623]: If we need this method?

File: src/models/backbones/hrnet.py [hunk 13]
```diff
@@ -0,0 +1,696 @@
...
+        modules = []
+        for i in range(num_modules):
+            modules.append(HighResolutionModule(
+                num_branches, block, num_blocks, num_inchannels, num_channels)
+            )
+            num_inchannels = modules[-1].get_num_inchannels()
+
+        return nn.Sequential(*modules), num_inchannels
+
+    def __stages(self, x: Tensor) -> List[Tensor]:
```
Comment (PososikTeam) [inline src/models/backbones/hrnet.py:597]: Please describe all parameters and write documentation

File: src/models/backbones/hrnet.py [hunk 14]
```diff
@@ -0,0 +1,758 @@
...
+
+        xl = [t(yl[-1]) if not isinstance(t, nn.Identity) else yl[i] for i, t in enumerate(self.transition3)]
+        yl = self.stage4(xl)
+        return yl
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Forward method.
+
+        Args:
+            x: Input tensor.
```
Comment (VladVin) [inline src/models/backbones/hrnet.py:676]: Not clear what shape is in input and on output

File: tests/models/backbones/test_hrnet.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+    def test_outputs_equals(self):
+        x = self.backbone(self._input)
+        x = self.neck(x)
+        x = F.avg_pool2d(x, kernel_size=x.size()[2:]).view(x.size(0), -1)
+        x = self.head(x)
+        self.assertTupleEqual(x.shape, (1, 10))
+
+
+class TestHRNetClassification_W48(TestHRNetClassification):
```
Comment (VladVin) [inline tests/models/backbones/test_hrnet.py:60]: I think we should remove this test, because it requires a lot of memory and compute

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/models/backbones/hrnet.py (hunk 14 above)
Comment (VladislavPatrushev) [reply src/models/backbones/hrnet.py:676]: Change output type

File: src/models/backbones/hrnet.py [hunk 16]
```diff
@@ -0,0 +1,758 @@
...
+        return yl
+
+    def forward_backbone_features(self, x: Tensor) -> Tuple[List[Tensor], List[Tensor]]:
+        """Forward backbone features and input tensor.
+
+        Args:
+            x: Input tensor.
+        """
+        features = self.forward(x)
+        features = [x] + features
```
[the remark this answers is not in the document]
Comment (VladislavPatrushev) [reply src/models/backbones/hrnet.py:692]: To know the shape for upsample in the neck

**Pull request merged**